packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -65,11 +66,15 @@ export interface IPendingLocalState {
  *
  * It verifies that all the ops are acked, are received in the right order and batch information is correct.
  */
-export class PendingStateManager {
+export class PendingStateManager implements IDisposable {
     private readonly pendingStates = new Deque<IPendingState>();
     private readonly initialStates: Deque<IPendingState>;
     private readonly initialClientId: string | undefined;
     private readonly initialClientSeqNum: number;
+    private readonly disposeOnce = new Lazy<void>(() => {
+        this.initialStates.clear();
```

Comment:
I think the spirit of the issue was to allow DDSs to know that their local ops are rejected.
But looking at the structure, there is no way currently to achieve that. I.e. acks are coming through SharedObject.processCore(op, local = true, localOpMetadata). So if we would want nacks to flow in similar way, we would need to design and come up with new API.
Alternatively we can use newAckBasedPromise(), where it matters. I think that's likely better approach, as number of places where we want to communicate failure to complete DDS change is tiny (mostly consensus data structures).
@markfields - what do you think?

Reply:
The latter sounds like what already needs to happen (based on the function comments):

```
    /**
     * Promises that are waiting for an ack from the server before resolving should use this instead of new Promise.
     * It ensures that if something changes that will interrupt that ack (e.g. the FluidDataStoreRuntime disposes),
     * the Promise will reject.
     * If runtime is disposed when this call is made, executor is not run and promise is rejected right away.
     */
    protected async newAckBasedPromise<T>(
```

Do you have an example of where this was supposed to be used?

Reply:
I'd say we will use it on as needed bases with current design.